[SDL-0274] Add preferred FPS to VideoStreamingCapability

shiniwat opened the pull request:
Fixes #1553, #1860 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Ran SDLVideoStreamingCapabilitySpec

#### Core Tests
- Core 4.5 (w/o having preferredFPS parameter)
- Core 6.1 (w/ preferredFPS parameter)

Core version / branch / commit hash / module tested against: Commit to Core 6.1 is pending
HMI name / version / branch / commit hash / module tested against: Our own HMI w/ Video streaming capability enabled.

### Summary
This PR adds preferredFPS parameter to VideoStreamingCapability, and modified SDLStreamingVideoLifecycleManager accordingly. 

### Changelog
##### Breaking Changes
* none

##### Enhancements
* SDL-0274

##### Bug Fixes
* issue #1553

### Tasks Remaining:
- [sdl_core issue #3243]

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

joeljfischer commented:
Hi @shiniwat, this proposal has not been earmarked for a release, so this PR has cannot be merged until that release. At the earliest that would be October. I can keep this PR open if you wish, but be aware that it will not be reviewed at this time.

shiniwat commented:
Understood. Thanks for letting it me know.

yoooriii commented on SmartDeviceLink/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -389,7 +396,14 @@ - (void)didEnterStateVideoStreamStarting {
         // Apply customEncoderSettings here. Note that value from HMI (such as maxBitrate) will be overwritten by custom settings.
         for (id key in self.customEncoderSettings.keyEnumerator) {
-            self.videoEncoderSettings[key] = [self.customEncoderSettings valueForKey:key];
+            if ([(NSString *)key isEqualToString:(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate] == YES) {
```
One should not compare a bool value with YES or NO, especially with YES:
1. The method -isEqualToString: already produces a bool value (YES or NO) which can be used directly without comparing it one more time to YES.
2. Let's take a more sophisticated example:
const BOOL bValue = (BOOL)2; // true condition
if (bValue) { /* true condition works */ }
if (bValue == YES) { /* true condition does not work */ }

shiniwat replied:
Thanks for the review. Fixed as suggested.

shiniwat commented:
Because SDL 0323 implementation needs to handle preferredFPS and max bitrate in the same way, and preferredFPS requires this PR, I have added the SDL 0323 implementation in this PR at [this commit](https://github.com/smartdevicelink/sdl_ios/pull/1613/commits/a2a77b21ef90412ccfebc55c0371f8d3e7d5153f).

joeljfischer commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -404,6 +409,8 @@ - (void)didEnterStateVideoStreamStarting {
             weakSelf.preferredFormats = @[format];
             weakSelf.preferredResolutions = @[resolution];
+            weakSelf.videoEncoderSettings[(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate] = @(defaultFrameRate);
+            SDLLogD(@"SDLCarWindow#didEnterStateVideoStreamStarting preferredFPS is default value (capability is nil)");
```
```suggestion
```

joeljfischer commented:
@shiniwat Please link to the SDL Core and HMI PRs that were used to test this PR.

joeljfischer commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -30,6 +30,7 @@ - (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)prefe
     self.diagonalScreenSize = @(diagonalScreenSize);
     self.pixelPerInch = @(pixelPerInch);
     self.scale = @(scale);
+    self.preferredFPS = @(15);
```
This is incorrect, the initializer should be deprecated and replaced. Please generate the file using the RPC generator and the correct branch of the rpc_spec. You'll need to modify the existing file instead of replacing it with the new file, but it will ensure that the new pieces are created correctly.

shiniwat replied:
Because RPC generator did not exist when I did initially pull-requested (which was in March), SDLVideoStreamingCapability was manually edited as you pointed out. I have updated the code based on RPC generator's generated code. Thanks.

joeljfischer reviewed (changes requested):
I cannot review the video lifecycle manager piece until the Core PR is approved, but I have a few notes I saw when reviewing other pieces.

shiniwat commented:
> @shiniwat Please link to the SDL Core and HMI PRs that were used to test this PR.

Hi @joeljfischer, here are the links:
sdl_core PR : https://github.com/smartdevicelink/sdl_core/pull/3437 (I am going to fix CI build failures by the way).
sdl_hmi PR: https://github.com/smartdevicelink/sdl_hmi/pull/436

joeljfischer commented:
@shiniwat the Core PR does not implement both proposals that are implemented by this PR and therefore cannot be used to test this PR. Please create two separate PRs for the two proposals that you currently implement in this one PR.

shiniwat commented:
Hi @joeljfischer, I have reverted the implementation commit for SDL 0323. Thanks.

shiniwat commented:
Hi @NicoleYarroch, I thought all change requests are resolved a while back, but forgot to re-request the review. I made that right before, thanks.

NicoleYarroch reviewed (changes requested):
I requested a few changes to the code to fix breaking changes.

NicoleYarroch commented:
@shiniwat Please tag me when you are ready for Livio to re-review this PR. Thanks!

shiniwat commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -28,9 +28,10 @@ NS_ASSUME_NONNULL_BEGIN
  @return A SDLVideoStreamingCapability object
  */
-- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale;
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale preferredFPS:(uint32_t)preferredFPS;
```
That makes sense. Thanks for pointing it out.

shiniwat commented on SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m:
```diff
@@ -73,10 +76,11 @@
+        expect(testStruct.preferredFPS).to(beNil());
     });
 
     it(@"Should initialize correctly with initWithPreferredResolution:maxBitrate:supportedFormats:hapticDataSupported:diagonalScreenSize:pixelPerInch:scale", ^ {
```
Understood. Updated SDLVideoStreamingCapabilitySpec.m accordingly.

shiniwat commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -387,6 +388,10 @@ - (void)didEnterStateVideoStreamStarting {
             if (capability.maxBitrate != nil) {
                 weakSelf.videoEncoderSettings[(__bridge NSString *) kVTCompressionPropertyKey_AverageBitRate] = [[NSNumber alloc] initWithUnsignedLongLong:(capability.maxBitrate.unsignedLongLongValue * 1000)];
             }
+            NSUInteger  preferredFPS = (capability.preferredFPS.unsignedIntValue == 0) ? DefaultFrameRate : capability.preferredFPS.unsignedIntValue;
```
Removed the extra white space.

shiniwat commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -84,6 +85,12 @@ NS_ASSUME_NONNULL_BEGIN
 
+/**
+ *  Preferred frame rate (assuming the case where resolution is the same as preferredResolution)
+ *  Optional
```
Understood that this comments needs to be created by rpc_spec generator.

shiniwat commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -28,9 +28,10 @@ NS_ASSUME_NONNULL_BEGIN
  @param diagonalScreenSize The diagonal screen size in inches
  @param pixelPerInch The diagonal resolution in pixels divided by the diagonal screen size in inches
  @param scale The scaling factor the app should use to change the size of the projecting view
+ @param preferredFPS The preferred frame rate specified by head unit
```
Updated the comment accordingly.

shiniwat commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -387,6 +388,10 @@ - (void)didEnterStateVideoStreamStarting {
+            NSUInteger  preferredFPS = (capability.preferredFPS.unsignedIntValue == 0) ? DefaultFrameRate : capability.preferredFPS.unsignedIntValue;
+            SDLLogD(@"videoEncoderSettings = %@", weakSelf.videoEncoderSettings);
+            weakSelf.videoEncoderSettings[(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate] = @(preferredFPS);
+            SDLLogD(@"SDLCarWindow#didEnterStateVideoStreamStarting preferredFPS=%lu", (unsigned long)preferredFPS);
```
Removed the additional SDLLogD, and merged with existing one.

NicoleYarroch commented:
@shiniwat I see that you made my requested changes. Just checking in to see if this PR is ready for me to review. Thanks!

shiniwat commented:
Hi @NicoleYarroch,
Yes, I have made all changes you requested. I was wondering, however, why building unit test (SmartDeviceLinkTests) failed. While it's most unlikely relevant to my changes, I thought I should take a look prior to requesting the review.
On my dev box, I was running into another unit test error, which seems to be specific to XCode 12.3. Because I've been tied up with other tasks right now, I will take a look sometime later (in a day or two).

NicoleYarroch commented:
@shiniwat Ah yes, there are known issues with building the tests on Xcode v.12.3. I have created a branch (https://github.com/smartdevicelink/sdl_ios/branches) with a workaround called "bugfix/Xcode_12.3_tests_not_building" if you would like to run the tests (please don't commit the changes on that branch since we expect an update to Carthage to fix this bug).

shiniwat commented:
Hi @NicoleYarroch,
I have locally merged bugfix/Xcode_12.3_tests_not_building branch, and confirmed that solved the test builds failure. I believe this PR is ready for review, then. Thanks.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -30,7 +31,22 @@ NS_ASSUME_NONNULL_BEGIN
+ @param preferredFPS The preferred frame rate per second of the head unit
+ @return A SDLVideoStreamingCapability object
+ */
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale preferredFPS:(uint32_t)preferredFPS;
```
We are trying to use the inits generated by the RPC Spec generator script. That init allows us to pass a `nil` value for `preferredFPS`.

shiniwat replied:
Understood. I have updated SDLVideoStreamingCapability.h based on the generated code (updated preferredFPS relevant code only).

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -30,7 +31,22 @@ NS_ASSUME_NONNULL_BEGIN
-- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale;
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale __deprecated_msg("Use initWithPreferredResolution ...
+
+/**
+ Convenience init for creating a video streaming capability with all parameters.
+
+ @param preferredResolution The preferred resolution of a video stream for decoding and rendering on HMI
+ @param maxBitrate The maximum bitrate of video stream that is supported, in kbps
+ @param supportedFormats Detailed information on each format supported by this system, in its preferred order
+ @param hapticDataSupported True if the system can utilize the haptic spatial data from the source being streamed
+ @param diagonalScreenSize The diagonal screen size in inches
+ @param pixelPerInch The diagonal resolution in pixels divided by the diagonal screen size in inches
+ @param scale The scaling factor the app should use to change the size of the projecting view
+ @param preferredFPS The preferred frame rate per second of the head unit
+ @return A SDLVideoStreamingCapability object
+ */
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale preferredFPS:(uint32_t)preferredFPS;
```
```suggestion
/**
 * @param preferredResolution - preferredResolution
 * @param maxBitrate - maxBitrate
 * @param supportedFormats - supportedFormats
 * @param hapticSpatialDataSupported - hapticSpatialDataSupported
 * @param diagonalScreenSize - diagonalScreenSize
 * @param pixelPerInch - pixelPerInch
 * @param scale - scale
 * @param preferredFPS - preferredFPS
 * @return A SDLVideoStreamingCapability object
 */
- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(nullable NSNumber<SDLUInt> *)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticSpatialDataSupported:(nullable NSNumber<SDLBool> *)hapticSpatialDataSupported diagonalScreenSize:(nullable NSNumber<SDLFloat> *)diagonalScreenSize pixelPerInch:(nullable NSNumber<SDLFloat> *)pixelPerInch scale:(nullable NSNumber<SDLFloat> *)scale preferredFPS:(nullable NSNumber<SDLUInt> *)preferredFPS;
```

shiniwat replied:
Updated SDLVideoStreamingCapability.h based on the generated code as you suggested.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -17,7 +17,15 @@
+
+// deprecated method, and will be removed in the future
 - (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale {
+    self = [self initWithPreferredResolution:preferredResolution maxBitrate:maxBitrate supportedFormats:supportedFormats hapticDataSupported:hapticDataSupported diagonalScreenSize:diagonalScreenSize pixelPerInch:pixelPerInch scale:scale preferredFPS:DEFAULT_FRAME_RATE];
+
+    return self;
```
When the generated init is used (see comment I left in the header file) you can set `nil` for the `preferredFPS` so you don't need the `DEFAULT_FRAME_RATE` anymore: 

```
self = [self initWithPreferredResolution:preferredResolution maxBitrate:@(maxBitrate) supportedFormats:supportedFormats hapticSpatialDataSupported:@(hapticDataSupported) diagonalScreenSize:@(diagonalScreenSize) pixelPerInch:@(pixelPerInch) scale:@(scale) preferredFPS:nil];
```

shiniwat replied:
I have updated SDLVideoStreamingCapability.m based on the generated code as per suggestion.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -13,6 +13,7 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+#define DEFAULT_FRAME_RATE 15
```
Can you revert this back to the way you had it in the first review? I think it makes sense for the `SDLStreamingVideoLifecycleManager` to manage the logic for setting the default frame rate value if `preferredFPS` is `nil`.

shiniwat replied:
Understood. Updated accordingly.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -84,6 +100,12 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (nullable, strong, nonatomic) NSNumber<SDLFloat> *scale;
 
+/**
+ The preferred frame rate per second of the head unit. The mobile application / app library may take other factors into account that constrain the frame rate lower than this value, but it should not perform streaming at a higher frame rate than this value.
+ @since SDL 7.1
+ */
```
```suggestion
/**
 * The preferred frame rate per second of the head unit. The mobile application / app library may take other factors into account that constrain the frame rate lower than this value, but it should not perform streaming at a higher frame rate than this value.
 * {"num_min_value": 0, "num_max_value": 2147483647}
 *
 * @added in SmartDeviceLink 7.1.0
 */
```

NicoleYarroch replied:
I just updated the docs with the generated documentation for the parameter

shiniwat replied:
Updated SDLVideoStreamingCapability.h based on the generated code. Thanks for the review.

NicoleYarroch reviewed (changes requested):
I left a suggestion for the `SDLVideoStreamingCapability`  inits so `nil` can be set for `preferredFPS`

NicoleYarroch commented:
@shiniwat  Please tag me when you are ready for Livio to do a re-review. Thanks!

shiniwat commented:
Hi @NicoleYarroch, I have updated SDLVideoStreamingCapability (.h and .m) based on the generated code, and updated the callers (which include test code) accordingly. I believe it's ready for re-review. Thanks.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.h:
```diff
@@ -30,7 +30,21 @@ NS_ASSUME_NONNULL_BEGIN
  @return A SDLVideoStreamingCapability object
  */
-- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale;
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale __deprecated_msg("Use initWithPreferredResolution ...
```
This `init` is not implemented in the **.m** file (I think the new init was accidentally copied twice?). One of the test cases is failing due to this.

shiniwat replied:
Oops, I copied some `init` from generated SDLVideoStreamingCapability.m, but parameter set in generated method was different... Not sure why building the framework didn't fail on my XCode. My apologies.

NicoleYarroch commented on SmartDeviceLink/public/SDLVideoStreamingCapability.m:
```diff
@@ -17,20 +17,27 @@
 @implementation SDLVideoStreamingCapability
 
-- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale {
+// deprecated method, and will be removed in the future
+- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(nullable NSNumber<SDLUInt> *)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticSpatialDataSupported:(nullable NSNumber<SDLBool> *)hapticSpatialDataSupported diagonalScreenSize:(nullable NSNumber<SDLFloat> *)diagonalScreenSize pixelPerInch:(null ...
+    self = [self initWithPreferredResolution:preferredResolution maxBitrate:maxBitrate supportedFormats:supportedFormats hapticSpatialDataSupported:hapticSpatialDataSupported diagonalScreenSize:diagonalScreenSize pixelPerInch:pixelPerInch scale:scale preferredFPS:nil];
+
+    return self;
+}
```
```suggestion
- (instancetype)initWithPreferredResolution:(nullable SDLImageResolution *)preferredResolution maxBitrate:(int32_t)maxBitrate supportedFormats:(nullable NSArray<SDLVideoStreamingFormat *> *)supportedFormats hapticDataSupported:(BOOL)hapticDataSupported diagonalScreenSize:(float)diagonalScreenSize pixelPerInch:(float)pixelPerInch scale:(float)scale {
    return [self initWithPreferredResolution:preferredResolution maxBitrate:@(maxBitrate) supportedFormats:supportedFormats hapticSpatialDataSupported:@(hapticDataSupported) diagonalScreenSize:@(diagonalScreenSize) pixelPerInch:@(pixelPerInch) scale:@(scale) preferredFPS:nil];
}
```

shiniwat replied:
Fixed the init. Thanks.

NicoleYarroch commented on SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLVideoStreamingCapabilitySpec.m:
```diff
@@ -85,6 +89,20 @@
         expect(testStruct.diagonalScreenSize).to(equal(testDiagonalScreenSize));
         expect(testStruct.pixelPerInch).to(equal(testPixelPerInch));
         expect(testStruct.scale).to(equal(testScale));
+        expect(testStruct.preferredFPS).to(equal(testPreferredFPS));
```
```suggestion
        expect(testStruct.preferredFPS).to(beNil());
```

shiniwat replied:
Oops, thanks for catching this.

NicoleYarroch reviewed (changes requested):
I found an issue with one of the `SDLVideoStreamingCapability` inits that should be easily fixed. I also discovered something odd with comparing the `videoEncoderSettings` value with the `customEncoderSettings`, please let me know if you can reproduce it on your end since I'm not quite sure what is going on under the hood.

NicoleYarroch commented:
@shiniwat Can you please tag me when this PR is ready for review? Thanks!

shiniwat commented on SmartDeviceLink/private/SDLStreamingVideoLifecycleManager.m:
```diff
@@ -414,7 +417,14 @@ - (void)didEnterStateVideoStreamStarting {
-            self.videoEncoderSettings[key] = [self.customEncoderSettings valueForKey:key];
+            if ([(NSString *)key isEqualToString:(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate]) {
+                // do NOT override framerate if custom setting is higher than current setting.
+                if ([self.customEncoderSettings valueForKey:key] < self.videoEncoderSettings[key]) {
```
Well, you're right that comparison should be explicitly made with NSNumber. Thanks for catching.
